### PR TITLE
Itemizer

### PR DESCRIPTION
**Pull request by bgorven**
PR for talentappstore/sprints#3

### CODE REVIEW

File: docs/schemas/itemMeta.json [hunk 1]
```diff
@@ -0,0 +1,996 @@
...
+      "type": "object",
+      "properties": {
+        "type": {
+          "type": "string",
+          "enum": ["singleLine"]
+        },
+        "minSize": {
+          "description": "Min number of characters",
+          "type": "number",
+          "default": 1
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: So far I haven't been using default at all in the schemas. We don't have any known tooling that could make use of it, so would prefer to not start now (let me know if I've missed something!)

File: docs/schemas/itemMeta.json [hunk 2]
```diff
@@ -0,0 +1,996 @@
...
+      }
+    },
+
+    "tas-multiline": {
+      "$schema": "http://json-schema.org/draft-04/schema",
+      "id": "http://talentappstore.github.io/tas-tenant-apis/tas-multiline",
+      "title": "tas-multiline",
+      "description": "A multiline text question item",
+      "type": "object",
+      "properties": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: There's too much copy and paste here. Can we have single line and multi line share a definition here (e.g. minSize)?

File: docs/schemas/itemMeta.json [hunk 3]
```diff
@@ -0,0 +1,996 @@
...
+      "description": "A number question item",
+      "type": "object",
+      "properties": {
+        "type": {
+          "type": "string",
+          "enum": ["number"]
+        },
+        "maximum": {
+          "description": "Max number size",
+          "type": "number"
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: These (and lots of others) can be made type == integer.

File: docs/schemas/itemMeta.json [hunk 4]
```diff
@@ -0,0 +1,996 @@
...
+        },
+        "version": {
+          "description": "Previous version ID",
+          "type": "string"
+        },
+        "job": {
+          "description": "ID for the job associated with this item",
+          "type": "string"
+        },
+        "candidate": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: I don't see why candidate ID ever needs to be in here, you are always within the scope of a candidate (or application) when referring to a resume.

File: docs/schemas/itemMeta.json [hunk 5]
```diff
@@ -0,0 +1,996 @@
...
+      "id": "http://talentappstore.github.io/tas-tenant-apis/tas-resume",
+      "title": "tas-resume",
+      "description": "A resume question item",
+      "type": "object",
+      "properties": {
+        "type": {
+          "type": "string",
+          "enum": ["resume"]
+        },
+        "version": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: Why do we need this in the API?

File: docs/schemas/itemMeta.json [hunk 6]
```diff
@@ -0,0 +1,996 @@
...
+          "default": "No"
+        }
+      }
+    },
+
+    "tas-resume": {
+      "$schema": "http://json-schema.org/draft-04/schema",
+      "id": "http://talentappstore.github.io/tas-tenant-apis/tas-resume",
+      "title": "tas-resume",
+      "description": "A resume question item",
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: Why do we have a type for resume - its really just a document isn't it?

File: docs/schemas/itemMeta.json [hunk 7]
```diff
@@ -0,0 +1,996 @@
...
+          "description": "Allowable file mime types",
+          "type": "array",
+          "items": { "type": "string" }
+        },
+        "maxSize": {
+          "description": "Maximum size of the file in bytes",
+          "type": "number",
+          "default": 1000000
+        },
+        "ascpectRatio": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: Why do we need this? Should it be an enum?

Comment (abraae) [inline docs/schemas/itemMeta.json]: And its misspelt.

File: docs/schemas/itemMeta.json [hunk 8]
```diff
@@ -0,0 +1,996 @@
...
+          "type": "number"
+        },
+        "minHeight": {
+          "description": "Min height for this image",
+          "type": "number"
+        }
+      }
+    },
+
+    "tas-signature": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: What is a signature in this context?

File: docs/schemas/itemMeta.json [hunk 9]
```diff
@@ -0,0 +1,996 @@
...
+      "type": "object",
+      "properties": {
+        "type": {
+          "type": "string",
+          "enum": ["signature"]
+        }
+      }
+    },
+
+    "tas-document": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: See comments above re resume.

File: docs/schemas/itemMeta.json [hunk 10]
```diff
@@ -0,0 +1,996 @@
...
+      "properties": {
+        "type": {
+          "type": "string",
+          "enum": ["audio"]
+        },
+        "maxLength": {
+          "description": "Max length of recording allowed (in minutes)",
+          "type": "number"
+        },
+        "accept": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: Why is this here without any constraints?

File: docs/schemas/itemMeta.json [hunk 11]
```diff
@@ -0,0 +1,996 @@
...
+        },
+        "maxSize": {
+          "description": "Maximum size of the file in bytes",
+          "type": "number",
+          "default": 1000000
+        }
+      }
+    },
+
+    "tas-video": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: Is this for future expansion? We shouldn't add anything we don't know is needed yet, it just creates more likelihood our predictions will be wrong.

File: docs/schemas/itemMeta.json [hunk 12]
```diff
@@ -0,0 +1,996 @@
...
+        },
+        "maxSize": {
+          "description": "Maximum size of the file in bytes",
+          "type": "number",
+          "default": 1000000
+        }
+      }
+    },
+
+    "tas-avatar": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: Do we need this now?

File: docs/schemas/itemMeta.json [hunk 13]
```diff
@@ -0,0 +1,996 @@
...
+        },
+        "maxSize": {
+          "description": "Maximum size of the file in bytes",
+          "type": "number",
+          "default": 1000000
+        }
+      }
+    },
+
+    "tas-profile": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: I thought social profile data went into vcard?

File: docs/schemas/itemMeta.json [hunk 14]
```diff
@@ -0,0 +1,996 @@
...
+          "items": { "type": "string" }
+        },
+        "defaultNetwork": {
+          "description": "Default social network",
+          "type": "string"
+        }
+      }
+    },
+
+    "tas-url": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: Is this needed yet?

File: docs/schemas/itemMeta.json [hunk 15]
```diff
@@ -0,0 +1,996 @@
...
+        },
+        "maxSize": {
+          "description": "Max number of characters",
+          "type": "number",
+          "default": 400
+        }
+      }
+    },
+
+    "tas-phone": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: Should be in vcard shouldn't it?

File: docs/schemas/itemMeta.json [hunk 16]
```diff
@@ -0,0 +1,996 @@
...
+        },
+        "phoneTypes": {
+          "description": "List of phone types allowed",
+          "type": "array",
+          "items": { "type": "string" }
+        }
+      }
+    },
+
+    "tas-email": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: Do we need this now?

File: docs/schemas/itemMeta.json [hunk 17]
```diff
@@ -0,0 +1,996 @@
...
+            "properties": {
+              "id": { "type": "string" },
+              "displayText": { "type": "string" }
+            }
+          }
+        }
+      }
+    },
+
+    "tas-enumMulti": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: Copy and paste alert! Shouldn't this be merged with the previous LOV type question type? They seem to be virtually identical.

File: docs/schemas/itemMeta.json [hunk 18]
```diff
@@ -0,0 +1,996 @@
...
+          "type": "number"
+        },
+        "maxSelections": {
+          "description": "Maximum number of selections allowed",
+          "type": "number"
+        }
+      }
+    },
+
+    "tas-enumDetails": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: As above, should all these be one type?

File: docs/schemas/itemMeta.json [hunk 19]
```diff
@@ -0,0 +1,996 @@
...
+          "type": "number"
+        },
+        "maxSelections": {
+          "description": "Maximum number of selections allowed",
+          "type": "number"
+        }
+      }
+    },
+
+    "tas-geolocation": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: Do we need this yet?

File: docs/schemas/itemMeta.json [hunk 20]
```diff
@@ -0,0 +1,996 @@
...
+        "type": {
+          "type": "string",
+          "enum": ["sheet"]
+        },
+        "columns": {
+          "description": "List of columns",
+          "type": "array",
+          "items": {
+            "type": "object",
+            "properties": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: Not sure how to use this - I thought each row would contain N items ($ref)?

File: docs/schemas/itemMeta.json [hunk 21]
```diff
@@ -0,0 +1,996 @@
...
+              "type":        {"type": "string"},
+              "format":      {"type": "string"},
+              "description": {"type": "string"}
+            }
+          }
+        }
+      }
+    },
+
+    "tas-wysiwyg": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: Whats the formatting? Markdown? Do we need this yet?

File: docs/schemas/itemMeta.json [hunk 22]
```diff
@@ -0,0 +1,996 @@
...
+          "default": 1000
+        },
+        "encoding": {
+          "description": "Characer encoding that should be used for returned data",
+          "type": "string"
+        }
+      }
+    },
+
+    "tas-address": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: I thought this was vcard?

File: docs/schemas/itemMeta.json [hunk 23]
```diff
@@ -0,0 +1,996 @@
...
+      "id": "http://talentappstore.github.io/tas-tenant-apis/tas-bool",
+      "title": "tas-bool",
+      "description": "A boolean question item",
+      "type": "object",
+      "properties": {
+        "type": {
+          "type": "string",
+          "enum": ["bool"]
+        },
+        "trueString": {
```
Comment (bgorven) [inline docs/schemas/itemMeta.json]: This indicates what to display for each option. Sometimes you want the candidate to have the options "Yes" and "No", other times you might want "true" and "false", or you could want to show a check box that says "☑ I agree with all of the above".

File: docs/schemas/itemMeta.json [hunk 24]
```diff
@@ -0,0 +1,996 @@
...
+        },
+        "addressTypes": {
+          "description": "Types of address",
+          "type": "array",
+          "items": { "type": "string" }
+        }
+      }
+    },
+
+    "tas-employment": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: Do we need this yet?

File: docs/schemas/itemMeta.json [hunk 25]
```diff
@@ -0,0 +1,996 @@
...
+        },
+        "employmentTypes": {
+          "description": "Types of employement",
+          "type": "array",
+          "items" : { "type": "string" }
+        }
+      }
+    },
+
+    "tas-reference": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: What's this?

File: docs/schemas/itemMeta.json [hunk 26]
```diff
@@ -0,0 +1,996 @@
...
+        },
+        "certificateTypes": {
+          "description": "Types of certification",
+          "type": "array",
+          "items": { "type": "string" }
+        }
+      }
+    },
+
+    "tas-volunteer": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: Same.

File: docs/schemas/itemMeta.json [hunk 27]
```diff
@@ -0,0 +1,996 @@
...
+          "enum": ["volunteer"]
+        },
+        "maxNumber": {
+          "description": "Max number of records allowed",
+          "type": "number"
+        }
+      }
+    },
+
+    "tas-skill": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: Same.

File: docs/schemas/itemMeta.json [hunk 28]
```diff
@@ -0,0 +1,996 @@
...
+          "enum": ["skill"]
+        },
+        "maxNumber": {
+          "description": "Max number of skill allowed",
+          "type": "number"
+        }
+      }
+    },
+
+    "tas-education": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: Same.

File: docs/schemas/itemMeta.json [hunk 29]
```diff
@@ -0,0 +1,996 @@
...
+          "type": "string"
+        },
+        "newestTimeAllowed": {
+          "description": "Newest time that can be selected",
+          "type": "string"
+        }
+      }
+    },
+
+    "tas-datetime": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: Copy and paste alert. Probably these three should be merged into a single date-time type with some additional behaviour flags.

File: docs/schemas/itemMeta.json [hunk 30]
```diff
@@ -0,0 +1,996 @@
...
+          "type": "string"
+        },
+        "newestDateAllowed": {
+          "description": "Newest datetime that can be selected",
+          "type": "string"
+        }
+      }
+    },
+
+    "tas-salary": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: same

File: docs/schemas/itemMeta.json [hunk 31]
```diff
@@ -0,0 +1,996 @@
...
+        },
+        "maximumDigitals": {
+          "description": "Max number characters",
+          "type": "number",
+          "default": 15
+        }
+      }
+    },
+
+    "tas-info": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: What's this?

File: docs/schemas/itemMeta.json [hunk 32]
```diff
@@ -0,0 +1,996 @@
...
+          "type": "string"
+        },
+        "text": {
+          "description": "Text for info",
+          "type": "string"
+        }
+      }
+    },
+
+    "tas-heading": {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: This is a more grunty question we'll need to discuss - should this be a question (ugly) or should it be a higher level thing that contains a bunch of questions?

File: docs/schemas/itemMeta.json [hunk 33]
```diff
@@ -0,0 +1,996 @@
...
+        { "$ref": "#/definitions/tas-education" },
+        { "$ref": "#/definitions/tas-date" },
+        { "$ref": "#/definitions/tas-time" },
+        { "$ref": "#/definitions/tas-datetime" },
+        { "$ref": "#/definitions/tas-salary" },
+        { "$ref": "#/definitions/tas-info" },
+        { "$ref": "#/definitions/tas-heading" }
+      ]
+    },
+    {
```
Comment (abraae) [inline docs/schemas/itemMeta.json]: I would have expected a reference to the previous stuff in here?

File: docs/schemas/itemRaw.json [hunk 34]
```diff
@@ -0,0 +1,733 @@
...
+      "type": "object",
+      "properties": {
+        "value": {
+          "description": "Answer for this item",
+          "type": "string"
+        }
+      }
+    },
+
+    "tas-singleLineAnswer": {
```
Comment (abraae) [inline docs/schemas/itemRaw.json]: Why is type included here? This is a value, not metadata.

File: docs/schemas/itemRaw.json [hunk 35]
```diff
@@ -0,0 +1,733 @@
...
+               "type": "string",
+               "enum": ["multiline"]
+             }
+           }
+         }
+      ]
+    },
+
+    "tas-numberAnswer": {
+      "description": "Schema for number answer",
```
Comment (abraae) [inline docs/schemas/itemRaw.json]: schema? This is a value not metadata.

File: docs/schemas/itemRaw.json [hunk 36]
```diff
@@ -0,0 +1,733 @@
...
+      "properties": {
+        "type": {
+          "type": "string",
+          "enum": ["number"]
+        },
+        "value": {
+          "description": "Answer for this item",
+          "type": "number",
+          "multipleOf" : 1,
+          "minimum": -2147483648,
```
Comment (abraae) [inline docs/schemas/itemRaw.json]: These look like metadata?

File: docs/schemas/itemRaw.json [hunk 37]
```diff
@@ -0,0 +1,733 @@
...
+            "format": {
+              "description": "Mime type of file",
+              "type": "string"
+            }
+          }
+        }
+      ]
+    },
+
+    "tas-signatureAnswer": {
```
Comment (abraae) [inline docs/schemas/itemRaw.json]: Aha now I see what signature is!

File: docs/schemas/itemMeta.json (hunk 31 above)
Comment (bgorven) [inline docs/schemas/itemMeta.json]: This is for if you've got a piece of information to add in amongst the questions in the form, that doesn't require an answer.

File: docs/schemas/itemMeta.json (hunk 32 above)
Comment (bgorven) [inline docs/schemas/itemMeta.json]: Sounds good to me

File: docs/schemas/itemMeta.json (hunk 18 above)
Comment (bgorven) [inline docs/schemas/itemMeta.json]: Sounds good. We basically were only using one of them anyway.

File: docs/schemas/itemRaw.json (hunk 34 above)
Comment (bgorven) [inline docs/schemas/itemRaw.json]: The examples don't validate without the type field, as the objects can fit multiple branches of the oneOf statement. Similarly, you can't deserialize the answers into the appropriate strongly typed objects unless there is something to tell the deserializer what type of object it's dealing with.

File: docs/schemas/itemRaw.json (hunk 35 above)
Comment (bgorven) [inline docs/schemas/itemRaw.json]: I've fixed the descriptions.

File: docs/schemas/itemRaw.json (hunk 36 above)
Comment (bgorven) [inline docs/schemas/itemRaw.json]: These constraints on the value ensure that the data will fit into a 32-bit integer.

File: docs/schemas/itemRaw.json (hunk 37 above)
Comment (bgorven) [inline docs/schemas/itemRaw.json]: I've removed it for now; we can always get it back from the commit history when we find a use for it.
